Sources/PT.PM.Common/Reflection/ReflectionCache.cs

```diff
@@ -1,53 +1,173 @@
-            new Lazy<Dictionary<string, Type>>(() =>
+        static ReflectionCache()
+        {
+            registeredTypes = new HashSet<Type>();
+            kindClassType = new Dictionary<string, Type>();
+            typeSerializableProperties = new Dictionary<Type, (byte NodeType, PropertyInfo[] Properties)>();
+
+            Assembly ustAssembly = typeof(Ust).Assembly;
+            var types = (UstType[]) Enum.GetValues(typeof(UstType));
+            ustTypes = new Type[types.Length];
+            var assemblyTypes = ustAssembly.GetTypes().Where(type => type.IsSubclassOf(typeof(Ust))).ToList();
+
+            foreach (UstType type in types)
             {
-                var result = new Dictionary<string, Type>();
-                foreach (Assembly assembly in AppDomain.CurrentDomain.GetAssemblies())
+                string typeStr = type.ToString();
+                Type ustType = null;
+                foreach (Type assemblyType in assemblyTypes)
                 {
-                    if (assembly.IsActual())
+                    if (assemblyType.Name.Equals(typeStr))
                     {
-                        foreach (Type type in assembly.GetTypes()
-                            .Where(myType => myType.IsClass && !myType.IsAbstract &&
-                            myType.GetInterfaces().Contains(typeof(IUst))))
-                        {
-                            result.Add(type.Name.ToLowerInvariant(), type);
-                        }
+                        ustType = assemblyType;
                     }
                 }
-                return result;
-            });
 
-        public static bool TryGetClassType(string kind, out Type type)
-        {
-            return UstKindClassType.Value.TryGetValue(kind.ToLowerInvariant(), out type);
+                if (ustType == null)
+                {
+                    throw new InvalidDataContractException($"NodeType {type} does not have corresponding Ust");
+                }
+
+                ustTypes[(int) type] = ustType;
+            }
+
+            RegisterTypes(typeof(Ust), true);
         }
 
-        public static PropertyInfo[] GetReadWriteClassProperties(this Type objectType)
+        public static void RegisterTypes(Type baseType, bool messagePack)
         {
-            if (!ustNodeProperties.TryGetValue(objectType, out PropertyInfo[] result))
+            lock (registeredTypes)
             {
-                result = objectType.GetProperties(BindingFlags.Public | BindingFlags.Instance)
-                    .Where(prop => prop.CanWrite && prop.CanRead &&
-                                   (prop.GetCustomAttribute<JsonIgnoreAttribute>() == null || prop.Name == nameof(Ust.TextSpans)))
-                    .ToArray();
-                ustNodeProperties.TryAdd(objectType, result);
+                if (registeredTypes.Contains(baseType))
+                {
+                    return;
+                }
+
+                registeredTypes.Add(baseType);
```

Comment:
Strictly speaking this is not sufficient protection against data races in general case. For instance, one thread adds hierarchy starting with `Ust` and another starting with `Expression` before the first thread have added `Expression` to the dictionary.

If such situation could never happen with any types we need a comment explaining why this is the case.

Reply:
Actually, I don't like the current implementation, because I want to get rid of situations with double addition. I will add a comment.